app/src/integrationTest/kotlin/maru/app/MaruFollowerTest.kt

```diff
@@ -342,35 +400,48 @@ class MaruFollowerTest {
         )
       }
     }
-    checkNetworkStackBlocksProduced(validatorStack, 2 * blocksToProduce)
-    checkNetworkStackBlocksProduced(followerStack, blocksToProduce)
+    checkNetworkStacksBlocksProduced(2 * blocksToProduce + residueBlocks, validatorStack)
+    checkNetworkStacksBlocksProduced(blocksToProduce, followerStack)
+    // ensure that the head of follower is at blocksToProduce
+    assertThat(followerStack.besuNode.getBlockNumber()).isEqualTo(blocksToProduce)
+
     peers.forEach {
       followerP2PNetwork.addPeer("${it.address}/p2p/${it.nodeId}")
     }
-    checkNetworkStackBlocksProduced(followerStack, 2 * blocksToProduce)
-  }
-
-  private fun checkValidatorAndFollowerBlocks(blocksToProduce: Int) {
-    await
-      .pollDelay(100.milliseconds.toJavaDuration())
-      .timeout(30.seconds.toJavaDuration())
-      .untilAsserted {
-        val blocksProducedByQbftValidator = validatorStack.besuNode.getMinedBlocks(blocksToProduce)
-        val blocksImportedByFollower = followerStack.besuNode.getMinedBlocks(blocksToProduce)
-        assertThat(blocksImportedByFollower).isEqualTo(blocksProducedByQbftValidator)
+    when (syncingConfig.syncTargetSelection) {
+      is SyncingConfig.SyncTargetSelection.Highest ->
+        checkValidatorAndFollowerBlocks(
+          2 * blocksToProduce + residueBlocks,
+        )
+      is SyncingConfig.SyncTargetSelection.MostFrequent -> {
+        checkValidatorAndFollowerBlocks(2 * blocksToProduce)
+        // ensure that the head of follower is at 2 * blocksToProduce
+        assertThat(followerStack.besuNode.getBlockNumber()).isEqualTo(2 * blocksToProduce)
       }
+    }
   }
 
-  private fun checkNetworkStackBlocksProduced(
-    stack: PeeringNodeNetworkStack,
+  private fun checkValidatorAndFollowerBlocks(blocksToProduce: Int) =
+    checkNetworkStacksBlocksProduced(blocksToProduce, validatorStack, followerStack)
+
+  private fun checkNetworkStacksBlocksProduced(
     blocksProduced: Int,
+    vararg stacks: PeeringNodeNetworkStack,
   ) {
     await
       .pollDelay(100.milliseconds.toJavaDuration())
       .timeout(30.seconds.toJavaDuration())
       .untilAsserted {
-        val blocksOnStack = stack.besuNode.getMinedBlocks(blocksProduced)
-        assertThat(blocksOnStack.size).isEqualTo(blocksProduced)
```

Comment:
This was wrong because `stack.besuNode.getMinedBlocks(blocksProduced)` would always return a list of blocks with size equal to `blocksProduced`, and put null on those that were failed to get, therefore I made changes here: https://github.com/Consensys/maru/pull/312/files#:~:text=%7D.mapNotNull%20%7B%20it.get().block%20%7D